File: csharp/v7/TypeScript/CSharpParserBase.ts
import { Parser, TokenStream } from "antlr4";
import CSharpParser, { Local_variable_declarationContext } from "./CSharpParser.js";

const ALL_SEMANTIC_FUNCTIONS = ["IsLocalVariableDeclaration"];

function parseNoSemantics(args: string[]): Set<string> {
    const result = new Set<string>();
    for (const a of args) {
        if (a.toLowerCase().startsWith("--no-semantics")) {
            const eq = a.indexOf('=');
            if (eq === -1) {
                for (const f of ALL_SEMANTIC_FUNCTIONS) result.add(f);
            } else {
                for (const f of a.substring(eq + 1).split(',')) result.add(f.trim());
            }
        }
    }
    return result;
}

export default abstract class CSharpParserBase extends Parser {
    private _noSemantics: Set<string>;

    constructor(input: TokenStream) {
        super(input);
        this._noSemantics = parseNoSemantics(process.argv);
    }

    protected IsRightArrow(): boolean { return this.areAdjacent(); }
    protected IsRightShift(): boolean { return this.areAdjacent(); }
    protected IsRightShiftAssignment(): boolean { return this.areAdjacent(); }

    private areAdjacent(): boolean {
        const first  = this._input.LT(-2);
        const second = this._input.LT(-1);
        return first !== null && second !== null &&
               first.tokenIndex + 1 === second.tokenIndex;
    }

    protected IsLocalVariableDeclaration(): boolean {
        if (this._noSemantics.has("IsLocalVariableDeclaration")) return true;
        const local_var_decl = this._ctx as Local_variable_declarationContext | null;
        if (local_var_decl == null) return true;
        const local_variable_type = local_var_decl.local_variable_type();
        if (local_variable_type == null) return true;
        if (local_variable_type.getText() === "var") return false;
        return true;
    }
}

File: csharp/v7/Java/CSharpParserBase.java
import org.antlr.v4.runtime.*;
import java.util.HashSet;
import java.util.Set;

public abstract class CSharpParserBase extends Parser
{
    private static final String[] ALL_SEMANTIC_FUNCTIONS = {
        "IsLocalVariableDeclaration"
    };

    private final Set<String> noSemantics;

    protected CSharpParserBase(TokenStream input)
    {
        super(input);
        noSemantics = parseNoSemantics(System.getProperty("sun.java.command", "").split("\\s+"));
    }

    private static Set<String> parseNoSemantics(String[] args)
    {
        Set<String> result = new HashSet<>();
        for (String a : args)
        {
            if (a.toLowerCase().startsWith("--no-semantics"))
            {
                int eq = a.indexOf('=');
                if (eq == -1)
                {
                    for (String f : ALL_SEMANTIC_FUNCTIONS) result.add(f);
                }
                else
                {
                    for (String f : a.substring(eq + 1).split(","))
                        result.add(f.trim());
                }
            }
        }
        return result;
    }

    protected boolean IsRightArrow() { return areAdjacent(); }
    protected boolean IsRightShift() { return areAdjacent(); }
    protected boolean IsRightShiftAssignment() { return areAdjacent(); }

    private boolean areAdjacent()
    {
        Token first  = _input.LT(-2);
        Token second = _input.LT(-1);
        return first != null && second != null &&
               first.getTokenIndex() + 1 == second.getTokenIndex();
    }

    protected boolean IsLocalVariableDeclaration()
    {
        if (noSemantics.contains("IsLocalVariableDeclaration")) return true;
        if (!(this._ctx instanceof CSharpParser.Local_variable_declarationContext)) {
            return false;
        }
        CSharpParser.Local_variable_declarationContext local_var_decl = (CSharpParser.Local_variable_declarationContext)this._ctx;
        if (local_var_decl == null) return true;
        CSharpParser.Local_variable_typeContext local_variable_type = local_var_decl.local_variable_type();
        if (local_variable_type == null) return true;
        if (local_variable_type.getText().equals("var")) return false;
        return true;
    }
}
